pkg/admission/validator/secret.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"context"
+	"fmt"
+
+	gcpvalidation "github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/validation"
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/gcp"
+
+	extensionswebhook "github.com/gardener/gardener/extensions/pkg/webhook"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type secret struct {
+	client client.Client
+}
+
+// NewSecretValidator returns a new instance of a secret validator.
+func NewSecretValidator() extensionswebhook.Validator {
+	return &secret{}
+}
+
+// InjectClient injects the given client into the validator.
+func (s *secret) InjectClient(client client.Client) error {
+	s.client = client
+	return nil
+}
+
+// Validate validates the given secret objects.
+func (s *secret) Validate(ctx context.Context, new, _ runtime.Object) error {
+	secret, ok := new.(*corev1.Secret)
+	if !ok {
+		return fmt.Errorf("wrong object type %T", new)
+	}
+
+	isUsed, err := isSecretUsedByShoot(ctx, s.client, secret, gcp.Type)
+	if err != nil {
+		return err
+	}
+
+	if isUsed {
+		if err := gcpvalidation.ValidateCloudProviderSecret(secret); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isSecretUsedByShoot(ctx context.Context, c client.Client, secret *corev1.Secret, providerType string) (bool, error) {
+	// controller-runtime cached client does not support non-exact field matches
+	secretBindings := &gardencorev1beta1.SecretBindingList{}
+	if err := c.List(ctx, secretBindings,
+		client.MatchingFields{"secretRef.namespace": secret.Namespace}); err != nil {
+		return false, err
+	}
+
+	for _, secretBinding := range secretBindings.Items {
+		// Filter out the the SecretBindings that do not reference the given secret
+		if secretBinding.SecretRef.Name != secret.Name {
```

Comment:
I think you can also compare the namespace names, am I right?
```suggestion
		if secretBinding.SecretRef.Name != secret.Name || secretBinding.SecretRef.Namespace != secret.Namespace {
```

Reply:
These secretBindings are already fetched with `client.MatchingFields{"secretRef.namespace": secret.Namespace}`. There is no need to check one more time that the secret namespace is matching the secretBinging secretRef.namespace. (Actually this if is needed because the cached client does support only 1 field selector :( )

Reply:
> (Actually this if is needed because the cached client does support only 1 field selector :( )

Does it make sense to add a TODO and come back later when the controller-runtime supports multiple?

Reply:
> These secretBindings are already fetched with client.MatchingFields{"secretRef.namespace": secret.Namespace}. 

You are right, I failed to see it :smile: